File: VB/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports Ex_2.Models
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc

Namespace Example.Controllers

	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult

			Return View()
		End Function

		Public Function GridViewPartial() As ActionResult
			Dim model = DataProvider.GetGridData()
			Return PartialView(model)
		End Function
		Public Function GridLookupPartial(ByVal CurrentID? As Integer) As ActionResult
			ViewData("Tags") = DataProvider.GetTags()
			Dim model As New GridDataItem() With {.ID = -1, .TagIDs = New Integer(){}}
			If CurrentID > -1 Then
				model = DataProvider.GetGridData().Where(Function(item) item.ID.Equals(CurrentID)).FirstOrDefault()
			End If

			Return PartialView(model)
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function GridViewUpdatePartial(ByVal item As GridDataItem) As ActionResult
			Dim model = DataProvider.GetGridData()
			If ModelState.IsValid Then
				Try
					DataProvider.UpdateGrid(item)
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			Else
				ViewData("EditError") = "Please, correct all errors."
			End If
			Return PartialView("GridViewPartial", model)
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function GridViewInsertPartial(ByVal item As GridDataItem) As ActionResult
			Dim model = DataProvider.GetGridData()
			If ModelState.IsValid Then
				Try
					DataProvider.InsertGrid(item)
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			Else
				ViewData("EditError") = "Please, correct all errors."
			End If
			Return PartialView("GridViewPartial", model)
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function GridViewDeletePartial(ByVal ID As System.Int32) As ActionResult
			Dim model = DataProvider.GetGridData()

			If ID >= 0 Then
				Try
					DataProvider.DeleteGrid(model.Where(Function(x) x.ID = ID).FirstOrDefault())
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			End If
			Return PartialView("GridViewPartial", model)
		End Function
	End Class
End Namespace

File: CS/Controllers/HomeController.cs
using Ex_2.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace Example.Controllers
{

    public class HomeController : Controller
    {
        public ActionResult Index()
        {

            return View();
        }

        public ActionResult GridViewPartial()
        {
            var model = DataProvider.GetGridData();
            return PartialView(model);
        }
        public ActionResult GridLookupPartial(int? CurrentID)
        {
            ViewData["Tags"] = DataProvider.GetTags();
            GridDataItem model = new GridDataItem() { ID = -1, TagIDs = new int[0] };
            if (CurrentID > -1)
            {
                model = DataProvider.GetGridData().Where(item => item.ID == CurrentID).FirstOrDefault();
            }
          
            return PartialView(model);
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult GridViewUpdatePartial(GridDataItem item)
        {
            var model = DataProvider.GetGridData();
            if (ModelState.IsValid)
            {
                try
                {
                    DataProvider.UpdateGrid(item);
                }
                catch (Exception e)
                {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";
            return PartialView("GridViewPartial", model);
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult GridViewInsertPartial(GridDataItem item)
        {
            var model = DataProvider.GetGridData();
            if (ModelState.IsValid)
            {
                try
                {
                    DataProvider.InsertGrid(item);
                }
                catch (Exception e)
                {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";
            return PartialView("GridViewPartial", model);
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult GridViewDeletePartial(System.Int32 ID)
        {
            var model = DataProvider.GetGridData();

            if (ID >= 0)
            {
                try
                {
                    DataProvider.DeleteGrid(model.Where(x => x.ID == ID).FirstOrDefault());
                }
                catch (Exception e)
                {
                    ViewData["EditError"] = e.Message;
                }
            }
            return PartialView("GridViewPartial", model);
        }
    }
}
